service/matching/physical_task_queue_manager.go

```diff
@@ -700,3 +714,55 @@ func (c *physicalTaskQueueManagerImpl) makePollerScalingDecisionImpl(
 		PollRequestDeltaSuggestion: delta,
 	}
 }
+
+func (c *physicalTaskQueueManagerImpl) getOrCreateTaskTracker(
+	intervals map[int32]*taskTracker,
+	priorityKey int32,
+) *taskTracker {
+	if priorityKey == 0 {
+		priorityKey = defaultPriorityLevel(c.config.PriorityLevels()) // assign default priority key
+	}
+
+	// First try with read lock for the common case where tracker already exists.
+	c.taskTrackerMu.RLock()
+	if tracker, ok := intervals[priorityKey]; ok {
+		c.taskTrackerMu.RUnlock()
+		return tracker
+	}
+	c.taskTrackerMu.RUnlock()
+
+	// Otherwise, we need to maybe create a new tracker with the write lock.
+	c.taskTrackerMu.Lock()
+	defer c.taskTrackerMu.Unlock()
+	if tracker, ok := intervals[priorityKey]; ok {
+		return tracker // tracker was created while we were waiting for the lock
+	}
+
+	// Initalize all task trackers together; or the timeframes won't line up.
+	c.tasksAddedInIntervals[priorityKey] = newTaskTracker(c.partitionMgr.engine.timeSource)
+	c.tasksDispatchedInIntervals[priorityKey] = newTaskTracker(c.partitionMgr.engine.timeSource)
```

Comment:
Took me more than an hour to figure this out; an unrelated test was failing and pointed me to this conclusion.